### ISSUE TITLE
systemd leaves around useless /var/tmp/systemd-private-* (for 30 days)

### ISSUE THREAD

**Issue by sourcejedi** [systemd/systemd#4401]
### Submission type
- [x ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> master
### Used distribution

> Fedora 24
### In case of bug report: Unexpected behaviour you saw

I see these do get cleaned (code below), but we pile up 30 days worth of them in `/var/tmp`.  That would explain why I have 42 of them, and that's only from two different services!

```
# Clear tmp directories separately, to make them easier to override
q /tmp 1777 root root 10d
q /var/tmp 1777 root root 30d

# Exclude namespace mountpoints created with PrivateTmp=yes
x /tmp/systemd-private-%b-*
X /tmp/systemd-private-%b-*/tmp
x /var/tmp/systemd-private-%b-*
X /var/tmp/systemd-private-%b-*/tmp
```

So, is it possible to insert something to clean the directories up more promptly?  They're not going to get re-used after a reboot.  Would something like the code below, still honour the exclusions?

```
R /tmp/systemd-private-*
R /var/tmp/systemd-private-%b-*
```

**Comment by poettering** [comment]
Yupp. I figure a patch like the one you suggested adding those R lines would make a lot of sense. 

Would be great if you could file a PR for adding that!

**Comment by sourcejedi** [comment]
I'll have a go! Thanks for looking at this proposal.

**Comment by sourcejedi** [comment]
I'm glad I tested this, because it didn't work :).

1. `R` lines are (deliberately) not affected by `x` lines.  This meant `systemd-tmpfiles --remove` would delete currently in-use mount points.

2. `R!` would make sure this only happened during sysinit.target (i.e. no removals after completing sysinit).  It might work ok, the services using PrivateTmp are not in sysinit.  I'm not thinking it's a great idea though.  sysinit services are expected to "specify all their dependencies manually". It's clear how this works for services which use directories created by systemd-tmpfiles. However I doubt anyone would remember, that `PrivateTmp` also implies an ordering dependency on `systemd-tmpfiles-setup`.

3. While testing (in a container) I was able to see mountpoints being removed during clean shutdown.  The mountpoints were left behind because my laptop crashes.  (Sometimes resume-from sleep fails, believed due to hardware age/damage).

=> alternative: do nothing, let the age-based cleanup handle it

4. I can't apply a cleanup age of 0 instead, because cleanup doesn't apply to any of the types that accept wildcards.

**Comment by poettering** [comment]
I am tempted to say that we should implicitly add a dependency on systemd-tmpfiles.service and the tmp.mount and var-tmp.mount units to all services that turn on PrivateTmp=, as they cannot really work properly otherwise. We should then document that in the man page, and we could then use the R! lines as suggested.

Note that these directories get cleaned up by PID 1 when the relevant service goes down. If that doesn't happen, then that's a bug in PID 1. If we assume that PID 1 doesn't suffer by the bug in that logic, then the existence of such a dir in /var/tmp can mean exactly two things: the service is currently running, or the system was abruptly turned off previously. In the latter case it's entirely safe to simply remove them without exception at next boot.

**Comment by mbiebl** [comment]
I do not remember any unclean shutdown, but have
```
drwx------   3 root    root        4096 Dez 27 10:00 systemd-private-8595abc09fcd40fa81df88f17ac1de6d-colord.service-ND9ILG
drwx------   3 root    root        4096 Dez 27 10:00 systemd-private-8595abc09fcd40fa81df88f17ac1de6d-rsyslog.service-KGsghB
drwx------   3 root    root        4096 Dez 27 10:00 systemd-private-8595abc09fcd40fa81df88f17ac1de6d-rtkit-daemon.service-iBtQvr
drwx------   3 root    root        4096 Dez 27 10:00 systemd-private-8595abc09fcd40fa81df88f17ac1de6d-systemd-timesyncd.service-WKPSWa
drwx------   3 root    root        4096 Dez 13 07:51 systemd-private-adea7c960eb8460e98282f7ee67743f5-systemd-hostnamed.service-AMN8Am
drwx------   3 root    root        4096 Dez 13 07:51 systemd-private-adea7c960eb8460e98282f7ee67743f5-systemd-timesyncd.service-HMN5o7
drwx------   3 root    root        4096 Dez 19 08:36 systemd-private-c102764191a245328174261df43ec542-colord.service-jiqPyn
drwx------   3 root    root        4096 Dez 19 08:36 systemd-private-c102764191a245328174261df43ec542-rsyslog.service-Ze5Mwv
drwx------   3 root    root        4096 Dez 19 08:36 systemd-private-c102764191a245328174261df43ec542-rtkit-daemon.service-ToIsu3
drwx------   3 root    root        4096 Dez 19 08:36 systemd-private-c102764191a245328174261df43ec542-systemd-timesyncd.service-KTz2N8
drwx------   3 root    root        4096 Nov 28 18:00 systemd-private-cb801acf280f41f9be9dfd5ab1dfc969-apache2.service-MZTarr
drwx------   3 root    root        4096 Nov 28 18:00 systemd-private-cb801acf280f41f9be9dfd5ab1dfc969-colord.service-PYPUye
drwx------   3 root    root        4096 Nov 28 18:00 systemd-private-cb801acf280f41f9be9dfd5ab1dfc969-rsyslog.service-qIFo9C
drwx------   3 root    root        4096 Nov 28 18:00 systemd-private-cb801acf280f41f9be9dfd5ab1dfc969-rtkit-daemon.service-FEBCmh
drwx------   3 root    root        4096 Nov 28 18:00 systemd-private-cb801acf280f41f9be9dfd5ab1dfc969-systemd-timesyncd.service-eKKJL2
```

**Comment by poettering** [comment]
Fix waiting in #4994.

**Comment by sourcejedi** [comment]
#4994 makes sense to me. To be explicit, my concern was that no-one would remember, when they needed to manually add a dependency on systemd-tmpfiles-setup.service. I see it can be added automatically (even with DefaultDependencies=no), and this fixes my concern.
